mutex_yield.c

```diff
@@ -0,0 +1,73 @@
+#include <libkern.h>
+#include <mutex_yield.h>
+#include <sched.h>
+#include <critical_section.h>
+#include <clock.h>
+
+void mtx_yield_init(mtx_yield_t *mtx) {
+  assert(mtx);
+
+  *mtx = 0;
+}
+void mtx_yield_lock(mtx_yield_t *mtx) {
+  assert(mtx);
+
+  while (true) {
+    while (*mtx == 1)
+      sched_yield(true);
+
+    cs_enter();
+    if (*mtx == 0) {
+      *mtx = 1;
+      cs_leave();
+      return;
+    }
+    cs_leave();
+  }
+}
+void mtx_yield_unlock(mtx_yield_t *mtx) {
+  assert(mtx);
+
+  cs_enter();
+  *mtx = 0;
```

Comment:
Well, this code already relies quite heavily on the fact that there is only a single core so most of the interesting stuff doesn't apply here.

The important things are that storing 0 in `mtx` is not interrupted and the memory accesses are not reordered by compiler (they can be reordered by the CPU since there is no other processor to see that). I am not very familiar with MIPS so this is something that should be checked, but I am pretty sure that `*mtx = 0` is already atomic (assuming mtx points to an aligned address, but it should, anyway). Obviously, using C11 atomics would be more explicit and safer. In that case `memory_order_relaxed` (i.e. the weakest one) is enough, since, as I mentioned earlier, there is no other CPU to see any reorderings.

As for compiler barriers, it is either `asm volatile("" : : : "memory")` – an empty inline assembly that tells GCC that memory was modified or `atomic_signal_fence` (again, the weakest one that only "synchronises the compiler") with `memory_order_seq_cst` to tell the compiler that nothing shall pass.